### ISSUE TITLE
New databases (features) and (primitives)

### ISSUE THREAD

**Issue by john-light** [john-light/softforks.org#5]
In his [opcodes zoo](https://gist.github.com/bigspider/2b0892da45884d26eb7f4c9cb2395a7d) document, Salvatore Ingala used a framework of analyzing soft forks by the "primitives" that they enable, and provided the following `opcode:primitives` groupings:

- CTV: send a precommitted transaction
- CSFS: sign a thing
- PC: commit to two things
- IKEY: push the taproot internal key on the stack
- CAT:
  - Concatenate two things
  - Vector commitments
  - Generic introspection
  - State-carrying UTXOs
- CCV: State-carrying UTXOs
- VAULT:
  - Multi-step transactions
  - State-carrying UTXOs
- TXHASH:
  - Generic introspection
  - State-carrying UTXOs
- APO: Replaceable state in layer 2s

This could be a useful additional dimension by which to compare the different soft forks. Concretely, I would like feedback on making the following changes:

1. Create a new database of `features` and move everything currently in the `keywords` column into this database. The `keywords` column would then be renamed `features` and become a [relation](https://www.notion.com/help/relations-and-rollups) like we have implemented for the `use cases` column. Each feature would have three properties: `bip`, `soft forks`, and `primitives`. `bip` is a link to the BIP for the given feature, if there is one e.g. `OP_CTV` would link to BIP-119. `soft forks` is a `relation` to the soft fork(s) that the `feature` is included in. `primitives` is a `relation` to the `primitives` that the feature enables.

2. Create another new database of `primitives` that contains all of the primitives that could be enabled by a given feature. Each primitive would have two properties: `features` and `description`. The `features` is a `relation` that can be used to assign the primitive to a given `feature` from the database in (1). For example, the feature `OP_CAT` would have the primitives `Concatenate two things`, `Vector commitments`, `Generic introspection`, and `State-carrying UTXOs`.

3. In the main `soft forks` database, include a new `primitives` column that is a [rollup](https://www.notion.com/help/relations-and-rollups) of all of the `primitives` enabled by a given soft fork according to the `features` that are included in the soft fork.

## Open questions

- What primitives should be included in the primitives database?
- Should we add subjective evaluations to the primitives, such as how Salvatore says some opcodes enable "badly-implemented" primitives? If so, should we surface those evaluations in the top-level soft forks database, and if so, how?

**Comment by bigspider** [comment]
Not sure how Notion works and if it's capable of representing the following, but in my mind I'd clearly separate the following 4 objects:
- `opcodes`
- `primitives` (or `features` perhaps): semantically meaningful building blocks that can be added to what's expressible in Script today. Like: vector commitments, concatenation, state-carrying UTXO, big numbers arithmetics, interactive fraud proofs for computations, zk-proofs, signatures, etc. They are the tools for wallet and smart contract developers.
- `applications` or use cases: things that can be built (or improved) with the primitives: vaults, rollups, Ark, sidechains, etc.
- `soft-forks`: combinations of opcodes (and/or other changes)

## Relations

The relations then go as follows:
- **opcodes** (or sometimes combinations of opcodes) can _implement a primitive_;
- **primitives** (or combinations of primitives) might _implement another primitives_;
- **primitives** (or combinations of primitives) might implement/improve an _application_.

## Comments

Now this all looks sound on paper, however: how well a set of opcodes implements a certain primitive; and how well a set of primitives implements a certain application _is one of the most important aspects_.

I think that the missing knowledge in the discourse is mainly:
- what applications are possible, and if/why they are compelling; more demos and prototypes are of course needed;
- the _relations_ (including the quality assessment) between opcodes/primitives/applications.

I might be wrong, but I'm a bit worried that any kind of database structure might focus most of the attention to the opcodes and primitives, rather than on the relations.
So far, all discourse has been mostly about opcodes and applications/soft-forks, making a discussion about relations impossible.

**Comment by john-light** [comment]
@bigspider Thanks for the suggestions! Can you say more about the "relations" quality that you mention? Perhaps by way of examples?

**Comment by bigspider** [comment]
> @bigspider Thanks for the suggestions! Can you say more about the "relations" quality that you mention? Perhaps by way of examples?

By the 'relations' I mean any _primitives implementing other primitives_, like:
- `concatenation` (+ `hashes`) implements `vector commitments`
- `state carrying` + `vector commitments` implements `interactive fraud proofs` (per my post on MATT)

or _primitives implementing applications_, for example:

- `fraud proofs` + `state-carrying` implements `optimistic rollups`

Then the same things should be done for _opcodes (or groups of opcodes) implement primitives_:
- `CAT` implements `concatenation` and `vector commitments` well; but also `state-carrying`, `commitments to a transaction`, etc. badly.
- `CSFS` implement `delegation` well
- `CCV` implements `state-carrying`well; but probably you can abuse the tweak to implement (badly) `vector commitments`, and probably combined with sighash flags it can implement (badly) `delegation`, etc.

Of course, there are various degrees of 'badly implementing', and that should also be underlined when possible.

**Comment by narula** [comment]
I think this is a useful framing. One thing I'm struggling with is that the table in notion contains `opcodes` and `softforks`, which are different things. One might say that a single `opcode` is its own soft-fork proposal, but 1) afaik no one is really advocating for just activating one `opcode` as a soft fork, that seems like a waste and 2) various combinations of opcodes should be captured and represented as distinct proposed `softforks`. 

There is some kind of exclusivity which is not captured yet -- "if you activated this opcode, you really wouldn't want to activate this one, they overlap too much." Or "these two opcodes solve the same problem (`primitive` in your parlance), but this one does it better for this `application` and this one does it better for this other `application`", and so on.

**Comment by john-light** [comment]
@narula 

> One thing I'm struggling with is that the table in notion contains `opcodes` and `softforks`, which are different things. One might say that a single `opcode` is its own soft-fork proposal, but 1) afaik no one is really advocating for just activating one `opcode` as a soft fork

Yeah the soft forks database started as the specific combinations of opcodes/BIPs that have been proposed for activation together e.g. Drivechain is described on drivechain.info as 2 BIPs (300+301), LNhance is 4 opcodes, etc. But then I thought about some opcodes that have also been proposed to be activated individually e.g. CTV and CAT, so I added those too.

> that seems like a waste

Personally I am happy to "take opcodes as they're ready" but am also aware that we already know of several combinations that are good so if it's not too much of a delay then I would prefer activating them together to minimize the logistical overhead. But I did want to acknowledge in the website that some opcodes are known to have benefits on their own and have therefore been considered for activation alone.

> 2) various combinations of opcodes should be captured and represented as distinct proposed softforks.

Currently we do show combinations in those cases where specific combinations have been proposed. I also acknowledge that it's blurry as to what should be considered a "soft fork" and what not, like should some of the opcode combinations that get thrown around like ["C3PO" and "C4"](https://xcancel.com/4moonsettler/status/1863000193831985652) be in the database even though they don't have a BIP or a website or something that formally says they should go together? I'm on the fence about it but lean to "yes" -- if it isn't trolling lol, there are a LOT of possible combinations of opcodes that I don't want to have to worry about tracking. But I do want to capture those good faith suggested combos and those cases where research shows when a combination of opcodes is "greater than the sum of its parts" i.e. some use case or efficiency gain or whatever is possible with combinations of opcodes that isn't possible with any of the individual opcodes on their own. 

> There is some kind of exclusivity which is not captured yet -- "if you activated this opcode, you really wouldn't want to activate this one, they overlap too much."

I agree this would also be useful to capture. The first thing that comes to mind is a new database for that, like a "Known bad combinations" database. WDYT? Or maybe that is too blunt of a tool for the job... can you think of another good way to represent that, either in the table/database paradigm of the current website or some other way?

> Or "these two opcodes solve the same problem (primitive in your parlance), but this one does it better for this application and this one does it better for this other application", and so on.

I also agree that this would be good to show. Open to ideas about how to implement this. One solution (another "blunt tool" suggestion) would be to create a new property on the (not-yet-existent) opcodes page in the (not-yet-existent) opcodes database called "Primitive usefulness score (1-10)" and simply write in a rating e.g. `Concatenation=10` if the opcode is the best way you could imagine implementing the primitive `Concatenation`. Then we can surface these on the opcodes and soft forks tables i.e. if an opcode is part of a soft fork then its `primitive usefulness score` will appear in the soft forks table, for ease of viewing and comparison.

**Comment by john-light** [comment]
There is now a proof of concept of this available for review here:
https://historical-meteor-e96.notion.site/18f47ba4d18d81a3b112fc7f47b57d84?v=18f47ba4d18d81f4b622000c699c7723

**Comment by john-light** [comment]
This feature is now live at bitcoin.softforks.org
